File: Aho/cmd/infix_to_postfix/infix_to_postfix.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"unicode"
)

var (
	lookahead int
	inExpr    expressionString_s
	outExpr   expressionString_s
)

func expr() {
	term()
	for {
		if lookahead == '+' {
			match('+')
			term()
			outExpr.appendTermToExprStr('+')
		} else if lookahead == '-' {
			match('-')
			term()
			outExpr.appendTermToExprStr('-')
		} else {
			return
		}
	}
}

func term() {
	if unicode.IsNumber(rune(lookahead)) {
		t := lookahead
		match(lookahead)
		outExpr.appendTermToExprStr(t)
	} else {
		fmt.Printf("syntax error in term()! %v - is not number!\n", string(rune(lookahead)))
		os.Exit(1)
	}
}

func match(t int) {
	if lookahead == t {
		_, lookahead = inExpr.getNextChar()
	} else {
		fmt.Println("syntax error in match()!")
		os.Exit(1)
	}
}

// func readOneRune(rd io.Reader) rune {
// in := bufio.NewReader(rd)
// foo, _, _ := in.ReadRune()
//
// return foo
// }

func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("read: ")
	text, _ := reader.ReadString('\n')

	if len([]rune(text)) > 1 {
		inExpr = buildExpressionString(text)
	} else {
		// Use default string
		inExpr = buildExpressionString("8-5+3+1-5")
	}

	outExpr = buildExpressionString("")

	_, lookahead = inExpr.getNextChar()

	expr()
	fmt.Println(outExpr.str)
}

File: Aho/infix_to_postfix/infix_to_postfix.js
let process = require("process")

class exprString_c {
    /**
     * 
     * @param {string} expr 
     */
    constructor(expr) {
        this.expression = expr
        this.pos = 0
    }

    getNextChar() {
        if ((this.pos + 1) > (this.expression.length)) {
            return [-1, ""]
        }

        let rt = this.expression[this.pos]
        this.pos = this.pos + 1
        return [this.pos, rt]
    }

    /**
     * 
     * @param {number} npos 
     * @returns {number}
     */
    setPos(npos) {
        if ((npos > this.expression.length) | (npos < 0)) {
            return -1
        }

        this.pos = npos
        return 0
    }

    /**
     * 
     * @param {string} nexpr 
     */
    loadNewExpression(nexpr) {
        this.expression = nexpr
        this.pos = 0
    }

    getExprLen() {
        return this.expression.length
    }

    /**
     * 
     * @param {string} ch 
     */
    appendChar(ch) {
        this.expression = this.expression + ch
    }
}

let lookahead = ""
let expressionLine = new exprString_c("9+5-2")
let expressionOut = new exprString_c("")

function expr() {
    term()
    while (true) {
        if (lookahead == "+") {
            match("+")
            term()
            expressionOut.appendChar("+")
        } else if (lookahead == "-") {
            match("-")
            term()
            expressionOut.appendChar("-")
        } else {
            return
        }
    }
}

function term() {
    if (!isNaN(lookahead)) {
        let t = lookahead
        match(lookahead)
        expressionOut.appendChar(t)
    } else {
        console.log("syntax error in term()! " + lookahead + "- is not number!")
        process.exit(0)
    }
}

/**
 * @param {string} t
 */
function match(t) {
    if (lookahead == t) {
        lookahead = expressionLine.getNextChar()[1]
    } else {
        console.log("syntax error in match()!")
        process.exit(0)
    }
}

(function main() {
    lookahead = expressionLine.getNextChar()[1]

    expr()

    console.log(expressionOut.expression)
}) ()
